File: 166-FractionToRecurringDecimal/166-FractionToRecurringDecimal.cpp
// Last updated: 04/04/2026, 23:36:18
class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        if(numerator==0) return "0";
        string result;
        if((numerator<0) ^ (denominator<0)) {
            result.push_back('-');
        }
        long long n=abs((long long)numerator);
        long long d=abs((long long)denominator);
        result += to_string(n/d);
        long long remainder=n%d;
        if(remainder==0) return result;
        result.push_back('.');
        unordered_map<long long,int> remMap;
        while(remainder!=0) {
            if(remMap.find(remainder)!=remMap.end()) {
                int idx=remMap[remainder];
                result.insert(idx,"(");
                result.push_back(')');
                break;
            }
            remMap[remainder]=result.length();
            remainder*=10;
            result.push_back('0'+remainder/d);
            remainder%=d;
        }
        return result;
    }
};

File: 166-FractionToRecurringDecimal/166-FractionToRecurringDecimal.java
// Last updated: 04/04/2026, 23:36:22
class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        if(numerator==0) return "0";
        StringBuilder ans = new StringBuilder();
        if((numerator<0)^(denominator<0)) {
            ans.append('-');
        }
        long n = Math.abs((long)numerator);
        long d = Math.abs((long)denominator);
        ans.append(n/d);
        Long rem = n%d;
        if(rem==0) {
            return ans.toString();
        }
        ans.append('.');
        HashMap<Long,Integer> map = new HashMap<>();
        while(rem!=0) {
            if(map.containsKey(rem)) {
                int idx = map.get(rem);
                ans.insert(idx,"(");
                ans.append(")");
                break;
            }
            map.put(rem,ans.length());
            rem *= 10;
            ans.append(rem/d);
            rem = rem%d;
        }
        return ans.toString();
    }
}
